spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Location.java

```diff
@@ -269,16 +292,50 @@ static LogicalLocation fromStackTraceElement(@NonNull StackTraceElement element)
             Objects.requireNonNull(bugInstance);
-            ClassAnnotation primaryClass = bugInstance.getPrimaryClass();
-            SourceLineAnnotation sourceLine = bugInstance.getPrimarySourceLineAnnotation();
-            return bugInstance.getAnnotations().stream().map(annotation -> {
-                String kind = findKind(annotation);
-                if (kind == null) {
-                    return null;
+            ClassAnnotation classAnnotation = null;
+            MethodAnnotation methodAnnotation = null;
+            FieldAnnotation fieldAnnotation = null;
+            LocalVariableAnnotation localVariableAnnotation = null;
+
+            for (BugAnnotation bugAnnotation : bugInstance.getAnnotations()) {
+                if (bugAnnotation instanceof ClassAnnotation) {
+                    classAnnotation = (ClassAnnotation) bugAnnotation;
+                } else if (bugAnnotation instanceof MethodAnnotation) {
+                    methodAnnotation = (MethodAnnotation) bugAnnotation;
+                } else if (bugAnnotation instanceof FieldAnnotation) {
+                    fieldAnnotation = (FieldAnnotation) bugAnnotation;
+                } else if (bugAnnotation instanceof LocalVariableAnnotation) {
+                    localVariableAnnotation = (LocalVariableAnnotation) bugAnnotation;
+                }
+            }
+
+            String fullyQualifiedName = "";
+            BugAnnotation annotation = null;
+            if (localVariableAnnotation != null) {
+                annotation = localVariableAnnotation;
+                fullyQualifiedName = String.format("%s#%s", methodAnnotation.getFullMethod(classAnnotation), localVariableAnnotation.getName());
+            } else {
+                if (fieldAnnotation != null) {
+                    annotation = fieldAnnotation;
+                    fullyQualifiedName = String.format("%s.%s", classAnnotation.getClassName(), fieldAnnotation.getFieldName());
+                } else {
+                    if (methodAnnotation != null) {
+                        annotation = methodAnnotation;
+                        fullyQualifiedName = methodAnnotation.getFullMethod(classAnnotation);
+                    } else if (classAnnotation != null) {
+                        annotation = classAnnotation;
+                        fullyQualifiedName = classAnnotation.getClassName();
+                    }
                 }
-                String name = annotation.format("givenClass", primaryClass);
-                return new LogicalLocation(name, null, kind, sourceLine.format("full", primaryClass), null);
-            }).filter(Objects::nonNull).findFirst();
+            }
+
+            if (annotation == null) {
+                return null;
+            }
+
+            String kind = findKind(annotation);
+            String name = annotation.format("givenClass", classAnnotation);
+            return Optional.of(new LogicalLocation(name, null, kind, fullyQualifiedName, null));
```

Comment:
I changed the logic to be more precise.
If the bugAnnotation has only class in the annotations, so that is the name
if the bugAnnotation has method, so that is the name
if the bugAnnotation has field, so it is class + field
if the bugAnnotation has variable, so it is method + field